File: src/Tiedragon.XmppMessenger.Core/Xmpp/XmppSaslScram.cs
using System.Security.Cryptography;
using System.Text;
using System.Xml.Linq;

namespace Tiedragon.XmppMessenger.Core.Xmpp;

public sealed class XmppSaslScram
{
    public const string MechanismSha1 = "SCRAM-SHA-1";

    public const string MechanismSha256 = "SCRAM-SHA-256";

    private readonly HashAlgorithmName _hashAlgorithm;
    private readonly string _username;
    private readonly string _password;
    private readonly string _clientNonce;
    private string? _clientFirstBare;
    private string? _serverFirstMessage;
    private byte[]? _serverSignature;

    public XmppSaslScram(string mechanism, string username, string password, string? clientNonce = null)
    {
        if (mechanism != MechanismSha1 && mechanism != MechanismSha256)
        {
            throw new ArgumentException("Unsupported SCRAM mechanism.", nameof(mechanism));
        }

        if (string.IsNullOrWhiteSpace(username))
        {
            throw new ArgumentException("Username is required.", nameof(username));
        }

        ArgumentNullException.ThrowIfNull(password);

        Mechanism = mechanism;
        _hashAlgorithm = mechanism == MechanismSha256 ? HashAlgorithmName.SHA256 : HashAlgorithmName.SHA1;
        _username = username;
        _password = password;
        _clientNonce = clientNonce ?? CreateNonce();
    }

    public string Mechanism { get; }

    public XElement CreateInitialAuthElement()
    {
        var initial = CreateClientFirstMessage();
        return new XElement(
            XName.Get("auth", XmppXmlNames.SaslNamespace),
            new XAttribute("mechanism", Mechanism),
            Convert.ToBase64String(Encoding.UTF8.GetBytes(initial)));
    }

    public XElement CreateResponseElement(string base64Challenge)
    {
        var challenge = Encoding.UTF8.GetString(Convert.FromBase64String(base64Challenge));
        var response = CreateClientFinalMessage(challenge);

        return new XElement(
            XName.Get("response", XmppXmlNames.SaslNamespace),
            Convert.ToBase64String(Encoding.UTF8.GetBytes(response)));
    }

    public static bool IsChallenge(XElement element)
    {
        ArgumentNullException.ThrowIfNull(element);
        return element.Name == XName.Get("challenge", XmppXmlNames.SaslNamespace);
    }

    public bool VerifyServerFinal(string base64ServerFinal)
    {
        if (_serverSignature is null)
        {
            throw new InvalidOperationException("SCRAM server signature is not available before client-final is created.");
        }

        var serverFinal = ParseAttributes(Encoding.UTF8.GetString(Convert.FromBase64String(base64ServerFinal)));
        return serverFinal.TryGetValue("v", out var verifier)
            && CryptographicOperations.FixedTimeEquals(Convert.FromBase64String(verifier), _serverSignature);
    }

    public string CreateClientFirstMessage()
    {
        _clientFirstBare = $"n={EscapeUsername(_username)},r={_clientNonce}";
        return $"n,,{_clientFirstBare}";
    }

    public string CreateClientFinalMessage(string serverFirstMessage)
    {
        if (_clientFirstBare is null)
        {
            CreateClientFirstMessage();
        }

        var attributes = ParseAttributes(serverFirstMessage);
        var nonce = Required(attributes, "r");
        if (!nonce.StartsWith(_clientNonce, StringComparison.Ordinal))
        {
            throw new FormatException("SCRAM server nonce does not extend the client nonce.");
        }

        var salt = Convert.FromBase64String(Required(attributes, "s"));
        var iterations = int.Parse(Required(attributes, "i"), System.Globalization.CultureInfo.InvariantCulture);
        if (iterations <= 0)
        {
            throw new FormatException("SCRAM iteration count must be positive.");
        }

        _serverFirstMessage = serverFirstMessage;
        var clientFinalWithoutProof = $"c=biws,r={nonce}";
        var authMessage = $"{_clientFirstBare},{_serverFirstMessage},{clientFinalWithoutProof}";

        var saltedPassword = Hi(_password, salt, iterations);
        var clientKey = Hmac(saltedPassword, "Client Key");
        var storedKey = Hash(clientKey);
        var clientSignature = Hmac(storedKey, authMessage);
        var clientProof = Xor(clientKey, clientSignature);
        var serverKey = Hmac(saltedPassword, "Server Key");
        _serverSignature = Hmac(serverKey, authMessage);

        return $"{clientFinalWithoutProof},p={Convert.ToBase64String(clientProof)}";
    }

    private static string CreateNonce()
    {
        return Convert.ToBase64String(RandomNumberGenerator.GetBytes(18)).TrimEnd('=');
    }

    private static string EscapeUsername(string username)
    {
        return username.Replace("=", "=3D", StringComparison.Ordinal)
            .Replace(",", "=2C", StringComparison.Ordinal);
    }

    private static string Required(IReadOnlyDictionary<string, string> attributes, string name)
    {
        return attributes.TryGetValue(name, out var value) && value.Length > 0
            ? value
            : throw new FormatException($"SCRAM attribute '{name}' is required.");
    }

    private static Dictionary<string, string> ParseAttributes(string message)
    {
        var result = new Dictionary<string, string>(StringComparer.Ordinal);
        foreach (var part in message.Split(',', StringSplitOptions.None))
        {
            var index = part.IndexOf('=');
            if (index <= 0)
            {
                throw new FormatException("SCRAM attribute is malformed.");
            }

            result[part[..index]] = part[(index + 1)..];
        }

        return result;
    }

    private byte[] Hi(string password, byte[] salt, int iterations)
    {
        return Rfc2898DeriveBytes.Pbkdf2(
            Encoding.UTF8.GetBytes(password),
            salt,
            iterations,
            _hashAlgorithm,
            _hashAlgorithm == HashAlgorithmName.SHA256 ? 32 : 20);
    }

    private byte[] Hash(byte[] value)
    {
        return _hashAlgorithm == HashAlgorithmName.SHA256
            ? SHA256.HashData(value)
            : SHA1.HashData(value);
    }

    private byte[] Hmac(byte[] key, string value)
    {
        var bytes = Encoding.UTF8.GetBytes(value);
        return _hashAlgorithm == HashAlgorithmName.SHA256
            ? HMACSHA256.HashData(key, bytes)
            : HMACSHA1.HashData(key, bytes);
    }

    private static byte[] Xor(byte[] left, byte[] right)
    {
        if (left.Length != right.Length)
        {
            throw new ArgumentException("SCRAM buffers must have equal length.");
        }

        var result = new byte[left.Length];
        for (var index = 0; index < left.Length; index++)
        {
            result[index] = (byte)(left[index] ^ right[index]);
        }

        return result;
    }
}

File: php/lib/Xmpp/XmppSaslScram.php
<?php
declare(strict_types=1);

namespace Tiedragon\Xmpp;

final class XmppSaslScram
{
    private string $clientNonce;
    private ?string $clientFirstBare = null;
    private ?string $serverFirstMessage = null;
    private ?string $serverSignature = null;

    public function __construct(
        public readonly string $mechanism,
        private readonly string $username,
        private readonly string $password,
        ?string $clientNonce = null
    ) {
        if (!in_array($mechanism, [XmppSasl::SCRAM_SHA_1, XmppSasl::SCRAM_SHA_256], true)) {
            throw new \InvalidArgumentException('Unsupported SCRAM mechanism.');
        }

        if (trim($username) === '') {
            throw new \InvalidArgumentException('SCRAM username is required.');
        }

        $this->clientNonce = $clientNonce ?: rtrim(base64_encode(random_bytes(18)), '=');
    }

    public function initialAuthElement(): string
    {
        return '<auth' . XmppXml::attributes([
            'xmlns' => XmppXml::SASL_NS,
            'mechanism' => $this->mechanism,
        ]) . '>' . XmppXml::escape(base64_encode($this->clientFirstMessage())) . '</auth>';
    }

    public function responseElementFromChallengeXml(string $challengeXml): string
    {
        return XmppSasl::response($this->clientFinalMessage($this->decodeSaslText($challengeXml)));
    }

    public function verifyServerFinalXml(string $serverFinalXml): bool
    {
        return $this->verifyServerFinal($this->decodeSaslText($serverFinalXml));
    }

    public function clientFirstMessage(): string
    {
        $this->clientFirstBare = 'n=' . self::escapeUsername($this->username) . ',r=' . $this->clientNonce;
        return 'n,,' . $this->clientFirstBare;
    }

    public function clientFinalMessage(string $serverFirstMessage): string
    {
        if ($this->clientFirstBare === null) {
            $this->clientFirstMessage();
        }

        $attributes = self::parseAttributes($serverFirstMessage);
        $nonce = self::required($attributes, 'r');
        if (!str_starts_with($nonce, $this->clientNonce)) {
            throw new \RuntimeException('SCRAM server nonce does not extend the client nonce.');
        }

        $salt = base64_decode(self::required($attributes, 's'), true);
        if ($salt === false) {
            throw new \RuntimeException('SCRAM salt is not valid base64.');
        }

        $iterations = (int)self::required($attributes, 'i');
        if ($iterations <= 0) {
            throw new \RuntimeException('SCRAM iteration count must be positive.');
        }

        $this->serverFirstMessage = $serverFirstMessage;
        $clientFinalWithoutProof = 'c=biws,r=' . $nonce;
        $authMessage = $this->clientFirstBare . ',' . $this->serverFirstMessage . ',' . $clientFinalWithoutProof;

        $saltedPassword = hash_pbkdf2($this->hashName(), $this->password, $salt, $iterations, $this->hashLength(), true);
        $clientKey = hash_hmac($this->hashName(), 'Client Key', $saltedPassword, true);
        $storedKey = hash($this->hashName(), $clientKey, true);
        $clientSignature = hash_hmac($this->hashName(), $authMessage, $storedKey, true);
        $clientProof = self::xorBytes($clientKey, $clientSignature);
        $serverKey = hash_hmac($this->hashName(), 'Server Key', $saltedPassword, true);
        $this->serverSignature = hash_hmac($this->hashName(), $authMessage, $serverKey, true);

        return $clientFinalWithoutProof . ',p=' . base64_encode($clientProof);
    }

    public function verifyServerFinal(string $serverFinalMessage): bool
    {
        if ($this->serverSignature === null) {
            throw new \RuntimeException('SCRAM server signature is not available before client-final is created.');
        }

        $attributes = self::parseAttributes($serverFinalMessage);
        $verifier = self::required($attributes, 'v');
        $decoded = base64_decode($verifier, true);
        return $decoded !== false && hash_equals($this->serverSignature, $decoded);
    }

    private function decodeSaslText(string $xml): string
    {
        $text = XmppSasl::text($xml);
        if ($text === '') {
            return '';
        }

        $decoded = base64_decode($text, true);
        if ($decoded === false) {
            throw new \RuntimeException('SASL SCRAM challenge is not valid base64.');
        }

        return $decoded;
    }

    private function hashName(): string
    {
        return $this->mechanism === XmppSasl::SCRAM_SHA_256 ? 'sha256' : 'sha1';
    }

    private function hashLength(): int
    {
        return $this->mechanism === XmppSasl::SCRAM_SHA_256 ? 32 : 20;
    }

    /**
     * @return array<string,string>
     */
    private static function parseAttributes(string $message): array
    {
        $result = [];
        foreach (explode(',', $message) as $part) {
            $index = strpos($part, '=');
            if ($index === false || $index <= 0) {
                throw new \RuntimeException('SCRAM attribute is malformed.');
            }
            $result[substr($part, 0, $index)] = substr($part, $index + 1);
        }

        return $result;
    }

    /**
     * @param array<string,string> $attributes
     */
    private static function required(array $attributes, string $name): string
    {
        if (!isset($attributes[$name]) || $attributes[$name] === '') {
            throw new \RuntimeException("SCRAM attribute '{$name}' is required.");
        }

        return $attributes[$name];
    }

    private static function escapeUsername(string $username): string
    {
        return str_replace(['=', ','], ['=3D', '=2C'], $username);
    }

    private static function xorBytes(string $left, string $right): string
    {
        if (strlen($left) !== strlen($right)) {
            throw new \RuntimeException('SCRAM buffers must have equal length.');
        }

        $result = '';
        $length = strlen($left);
        for ($i = 0; $i < $length; $i++) {
            $result .= $left[$i] ^ $right[$i];
        }

        return $result;
    }
}
